[Design]: Changes to improve performance

davidfowl opened the pull request:
- Track disposables in a list with lock instead of a concurrent bag
- Allocate a single _createServiceAccessor instead of one per instance
- Removed syncObject and just locked the _resolvedServices

Measurements on the way...

/cc @rynowak @halter73 @pranavkm @lodejard @benaadams

davidfowl committed:
Changes to improve performance
- Track disposables in a list with lock instead of a concurrent bag
- Allocate a single _createServiceAccessor instead of one per instance
- Removed syncObject and just locked the _resolvedServices

davidfowl committed:
Bleh

davidfowl commented on src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs:
```diff
@@ -145,25 +136,35 @@ internal IServiceCallSite GetResolveCallSite(IService service, ISet<Type> callSi
         private object CaptureDisposable(object service)
         {
+            // Skip capturing disposables for the root
+            if (_root == this)
```
Nobody disposes the root scope, an argument can be made that it is possible but the interface for the top level (IServiceProvider) requires a downcast to IDisposable to do so and that's pretty rare unless you're scoped. We can let the GC handle those objects.

benaadams commented:
This halves allocations for DI

![vs](http://aoa.blob.core.windows.net/aspnet/new-di.png)

![0](http://aoa.blob.core.windows.net/aspnet/new-di-0.png)

![1](http://aoa.blob.core.windows.net/aspnet/new-di-1.png)

rynowak commented on src/Microsoft.Framework.DependencyInjection/Internal/ConcurrentDictionaryExtensions.cs:
```diff
@@ -0,0 +1,43 @@
+                    return value;
+                }
+
+                value = valueFactory(key, arg);
```
You can make this code a little clearer by getting rid of the `while` and changing the `TryAdd` to `GetOrAdd(key, value)`. If your add fails, it's because someone was contending with you to also add a value and so you'll get the value they added.

pranavkm replied:
:+1:

davidfowl replied:
Feels like this is easier to read.

rynowak replied:
Follow your :heart: - but this gives me the impression that it will run an arbitrary number of times, but it will only ever run 1.5 times at most.

rynowak commented on src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs:
```diff
@@ -56,16 +47,16 @@ internal ServiceProvider(ServiceProvider parent)
         public object GetService(Type serviceType)
         {
-            var realizedService = _table.RealizedServices.GetOrAdd(serviceType, _createServiceAccessor);
+            var realizedService = _table.RealizedServices.GetOrAdd(serviceType, (type, sp) => CreateServiceAccessor(type, sp), this);
```
IMO worth having this in a static field and documenting why it is just so.

davidfowl replied:
I started that way ya know :smile:

rynowak commented on src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs:
```diff
@@ -56,16 +47,16 @@ internal ServiceProvider(ServiceProvider parent)
         }
 
-        private Func<ServiceProvider, object> CreateServiceAccessor(Type serviceType)
+        private static Func<ServiceProvider, object> CreateServiceAccessor(Type serviceType, ServiceProvider sp)
```
`sp`? horrid

rynowak commented on src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs:
```diff
@@ -145,25 +136,35 @@ internal IServiceCallSite GetResolveCallSite(IService service, ISet<Type> callSi
                 if (disposable != null)
                 {
-                    _disposables.Add(disposable);
+                    lock (_disposables)
```
Do you care about the race condition that arises when someone is trying to add a disposable **while** the dispose methods are being run?

davidfowl replied:
Not really, but we have no tests to verify it either. We don't even seem to check if we call GetService from a disposed service provider either.

rynowak commented:
overall looks good - main point here seems to be assuming that the service provider is low-contention and we can do simpler/lighter things because of that.

pranavkm commented on src/Microsoft.Framework.DependencyInjection/Internal/ConcurrentDictionaryExtensions.cs:
```diff
@@ -0,0 +1,43 @@
+        // This lets us pass a state parameter allocation free GetOrAdd
+        internal static TValue GetOrAdd<TKey, TValue, TArg>(this ConcurrentDictionary<TKey, TValue> dictionary, TKey key, Func<TKey, TArg, TValue> valueFactory, TArg arg)
+        {
+            if (dictionary == null)
```
If it's internal code, consider removing these exceptions and replacing it with `Debug.Assert` instead.

pranavkm commented on src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs:
```diff
@@ -145,25 +136,35 @@ internal IServiceCallSite GetResolveCallSite(IService service, ISet<Type> callSi
-            var disposables = Interlocked.Exchange(ref _disposables, null);
-
-            if (disposables != null)
+            lock (_disposables)
```
Isn't the guideline to lock on a `_disposablesLock`?

rynowak replied:
It's safe to lock on an arbitrary object when other code can't get at it. What you're suggesting is *safer* - but that's not the point. Your safety is not guaranteed.

pranavkm commented on src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs:
```diff
@@ -312,7 +313,7 @@ private static Expression Lock(Expression providerExpression, Expression body)
                 // The C# compiler would copy the lock object to guard against mutation.
                 // We don't, since we know the lock object is readonly.
-                var syncField = Expression.Field(providerExpression, "_sync");
+                var syncField = Expression.Field(providerExpression, "_resolvedServices");
```
Does `nameof(provider_resolvedServices)` work here?

davidfowl replied:
Yea, that would be better

davidfowl replied:
Don't have access to it.

benaadams commented:
788k -> 384k (-51%)

If you didn't want to go full pooling, but were ok with partial pooling, you could pool the dictionary and list and clear them up in the disposable; which would trim it to 144k but limit the scope for the pooling.

788k -> 144k (-82%)

Though likely with that you'd need to return to an object for locking and add bool for dispose check

davidfowl committed:
CR feedback

davidfowl commented:
updated. 

Going to try a few more things:
- Don't allocate disposables until required.

davidfowl committed:
Don't allocate disposable list until used
- Only use it for transient objects
- Use _resolvedServices as a lockObject to avoid allocating a lockObj

davidfowl commented:
@benaadams New changes for ya.

pranavkm commented:
Looks :shipit:  to me. Will wait for @benaadams's numbers though.

davidfowl committed:
CR Feedback

benaadams commented:
Improved:

![delayed-alloc](http://aoa.blob.core.windows.net/aspnet/di-scope-0.png)

And down overall for the module/dll:

![delayed-alloc](http://aoa.blob.core.windows.net/aspnet/di-scope-1.png)

davidfowl commented:
Merged
